ci: parse TEST_RUNNER_EXTRA into an array

m3dwards opened the pull request:
While working on CI I wanted to disable some functional tests so I used the `TEST_RUNNER_EXTRA` var. The problem I had was tests that have flags such as `rpc_bind.py --ipv6` must be passed in quotes otherwise the `--ipv6` portion will be considered an argument to `test_runner.py` rather than a test name.

This change allows proper parsing of quotes and complex values such as:

```shell
TEST_RUNNER_EXTRA='--exclude "rpc_bind.py --ipv6,feature_proxy.py"'
```

Update:

While testing this it was noticed that `test_runner.py` when given `--exclude "rpc_bind.py --ipv6"` will exclude all `rpc_bind.py` tests so this PR has been updated to include a change to the test runner to only exclude the specific test if you pass an arg or exclude all tests of that name if you do not pass an arg. `--exclude rpc_bind.py` will exclude all three variants and `--exclude rpc_bind --ipv6` will only exclude the IPV6 variant.

m3dwards committed:
ci: parse TEST_RUNNER_EXTRA into an array

Allows for parsing quotes and multiple arguments such as TEST_RUNNER_EXTRA='--exclude "rpc_bind.py --ipv6,feature_proxy.py"'

maflcko reviewed (approved):
I can't review this, because it is written in bash, but testing seems to indicate that this works as intended for some reason.

ACK b9c0b79ab7a0d4dafd5386a78f4fd747236c8f88

hebasto commented:
> This change allows proper parsing of quotes and complex values such as:
> 
> ```shell
> TEST_RUNNER_EXTRA='--exclude "rpc_bind.py --ipv6, feature_proxy.py"'
> ```

Does such an example work on Windows in Command Prompt and PowerShell?

hebasto commented:
> > Does such an example work on Windows in Command Prompt and PowerShell?
> 
> PowerShell and Command Prompt are not supported by `./ci/`. Only `bash`, according to https://github.com/bitcoin/bitcoin/tree/master/ci#running-a-stage-locally

I'm aware of that. I'm asking about `TEST_RUNNER_EXTRA` variable content given in the example.

m3dwards commented:
> Does such an example work on Windows in Command Prompt and PowerShell?

I will test it.

m3dwards commented on ci/test/03_test_script.sh:
```diff
@@ -161,7 +161,9 @@ fi
 
 if [ "$RUN_FUNCTIONAL_TESTS" = "true" ]; then
   # shellcheck disable=SC2086
```
The comment? I'm not sure

m3dwards commented:
Force pushed but only changed the commit description as it was a bad example with an extra space

m3dwards commented:
> Does such an example work on Windows in Command Prompt and PowerShell?

I can confirm that `TEST_RUNNER_EXTRA='--exclude "rpc_bind.py --ipv6,feature_proxy.py"'` works with cmd as defined in `.github/workflows/ci.yml`. Not tested Powershell but I don't believe we use this with Powershell.

[Windows CI run using that value]( https://github.com/m3dwards/bitcoin/actions/runs/9419238354/job/25948555814)

hebasto reviewed (commented):
Tested 3d976000afccac4e89931496f23cfe8593daeb75 on Ubuntu 24.04.

Passing `--exclude "rpc_bind.py --ipv6"` excludes not only the specified test but also `rpc_bind.py --ipv4` and `rpc_bind.py --nonloopback`.

m3dwards commented:
Converting back to draft to see if I can fix the bug @hebasto found.

m3dwards committed:
test: allow excluding func test by name and arg

Can now specify test_runner.py --exclude "rpc_bind.py --ipv6" and have only that test variant excluded

m3dwards commented:
Updated `test_runner.py` so that:

`--exclude "rpc_bind.py --IPV6`  will only exclude that variant.
`--exclude rpc_bind.py` will exclude all variants

Green test run with `TEST_RUNNER_EXTRA: --exclude "rpc_bind.py --ipv6, feature_proxy.py"`: https://github.com/m3dwards/bitcoin/actions/runs/9485948819/job/26139127116. Note that `rpc_bind.py --ipv4` and `rpc_bind.py --nonloopback` were both run but `rpc_bind.py --ipv6` and `feature_proxy.py` did not run.

Spaces between test names now also works.

hebasto reviewed (approved):
ACK 8131bf7483c0ea10d3573c9f2e977d19d8569b7f, tested on Ubuntu 23.10 and Windows 11.

In the command prompt on Windows, the following commands work as expected:
```
>set EXTRA=--exclude "rpc_bind.py --ipv6, feature_proxy.py"
>py -3 test\functional\test_runner.py %EXTRA%
```

which is sufficient for our CI if needed.

achow101 commented:
ACK 8131bf7483c0ea10d3573c9f2e977d19d8569b7f

m3dwards commented:
@maflcko no apology needed, apologies I introduced a bug!

I will work out what's going on and propose a fix PR ASAP.

Thanks for the logs.
